kernel/src/table_changes/physical_to_logical.rs

```diff
@@ -0,0 +1,191 @@
+#[allow(unused)]
+fn get_generated_columns(scan_file: &CdfScanFile) -> DeltaResult<HashMap<&str, Expression>> {
+    let timestamp = Scalar::timestamp_from_millis(scan_file.commit_timestamp)?;
+    let version = scan_file.commit_version;
+    let change_type: Expression = match scan_file.scan_type {
+        CdfScanFileType::Cdc => column_expr!("_change_type"),
+        CdfScanFileType::Add => "insert".into(),
+
+        CdfScanFileType::Remove => "delete".into(),
+    };
+    let expressions = [
+        ("_change_type", change_type),
+        ("_commit_version", Expression::literal(version)),
+        ("_commit_timestamp", timestamp.into()),
+    ];
+    Ok(expressions.into_iter().collect())
+}
+
+/// Generates the expression used to convert physical data from the `scan_file` path into logical
+/// data matching the `global_state.logical_schema`
+#[allow(unused)]
+fn get_expression(
+    scan_file: &CdfScanFile,
+    logical_schema: &StructType,
+    all_fields: &[ColumnType],
+) -> DeltaResult<Expression> {
+    let mut generated_columns = get_generated_columns(scan_file)?;
+    let all_fields = all_fields
+        .iter()
+        .map(|field| match field {
+            ColumnType::Partition(field_idx) => {
+                let field = logical_schema.fields.get_index(*field_idx);
+                let Some((_, field)) = field else {
+                    return Err(Error::generic(
+                        "logical schema did not contain expected field, can't transform data",
+                    ));
+                };
+                let name = field.physical_name();
+                let value_expression =
+                    parse_partition_value(scan_file.partition_values.get(name), field.data_type())?;
+                Ok(value_expression.into())
+            }
+            ColumnType::Selected(field_name) => {
+                // Remove to take ownership
+                let generated_column = generated_columns.remove(field_name.as_str());
+                Ok(generated_column.unwrap_or_else(|| ColumnName::new([field_name]).into()))
+            }
+        })
+        .try_collect()?;
+    Ok(Expression::Struct(all_fields))
+}
+
+/// Gets the physical schema that will be used to read data in the `scan_file` path.
+#[allow(unused)]
+fn get_read_schema(scan_file: &CdfScanFile, read_schema: &StructType) -> SchemaRef {
+    if scan_file.scan_type == CdfScanFileType::Cdc {
+        let change_type = StructField::new("_change_type", DataType::STRING, false);
+        let fields = read_schema.fields().cloned().chain(iter::once(change_type));
+        StructType::new(fields).into()
+    } else {
+        read_schema.clone().into()
+    }
+}
+
+/// Reads the data at the `resolved_scan_file` and transforms the data from physical to logical.
+/// The result is a fallible iterator of [`ScanResult`] containing the logical data.
+#[allow(unused)]
+pub(crate) fn read_scan_data(
```

Comment:
I think this may be in the wrong spot? I like that we have a `physical_to_logical` module. but this is really a high level API that (1) gets the physical-to-logical expression, (2) reads the parquet file in the scan file, and (3) evaluates the expression and bundles in ScanResult. Maybe just move it out of this module for now? could go in the top level mod.rs? this feels engine-ish as we are making some decisions on how to stitch our APIs together

Reply:
and if we keep this in this PR should probably add a few tests? (also one option is punting this to the execute PR)

Reply:
Will put in execute PR. 

In terms of testing, it's a lot of work to write a test that isn't just an E2E test since this requires real paths to read data.

Reply:
yea sg!